File: apps/mfe-evaluator/src/lib/evaluator/arabic-amount-words.ts
/** تفقيط مبالغ الريال السعودي لعرض «كتابة» بجانب الحقول الرقمية. */

const ONES = [
  "",
  "واحد",
  "اثنان",
  "ثلاثة",
  "أربعة",
  "خمسة",
  "ستة",
  "سبعة",
  "ثمانية",
  "تسعة",
  "عشرة",
  "أحد عشر",
  "اثنا عشر",
  "ثلاثة عشر",
  "أربعة عشر",
  "خمسة عشر",
  "ستة عشر",
  "سبعة عشر",
  "ثمانية عشر",
  "تسعة عشر",
];

const TENS = [
  "",
  "",
  "عشرون",
  "ثلاثون",
  "أربعون",
  "خمسون",
  "ستون",
  "سبعون",
  "ثمانون",
  "تسعون",
];

const HUNDREDS = [
  "",
  "مائة",
  "مائتان",
  "ثلاثمائة",
  "أربعمائة",
  "خمسمائة",
  "ستمائة",
  "سبعمائة",
  "ثمانمائة",
  "تسعمائة",
];

function underHundred(n: number): string {
  if (n < 20) return ONES[n] ?? "";
  const t = Math.floor(n / 10);
  const o = n % 10;
  if (!o) return TENS[t] ?? "";
  return `${ONES[o]} و${TENS[t]}`;
}

function underThousand(n: number): string {
  if (n < 100) return underHundred(n);
  const h = Math.floor(n / 100);
  const rest = n % 100;
  const head = HUNDREDS[h] ?? "";
  if (!rest) return head;
  return `${head} و${underHundred(rest)}`;
}

type Scale = {
  value: number;
  one: string;
  dual: string;
  few: string;
  many: string;
};

const SCALES: Scale[] = [
  { value: 1_000_000_000, one: "مليار", dual: "ملياران", few: "مليارات", many: "مليار" },
  { value: 1_000_000, one: "مليون", dual: "مليونان", few: "ملايين", many: "مليون" },
  { value: 1_000, one: "ألف", dual: "ألفان", few: "آلاف", many: "ألف" },
];

function scaleLabel(count: number, scale: Scale): string {
  if (count === 1) return scale.one;
  if (count === 2) return scale.dual;
  if (count >= 3 && count <= 10) return `${underThousand(count)} ${scale.few}`;
  return `${underThousand(count)} ${scale.many}`;
}

function integerToArabicWords(n: number): string {
  if (n === 0) return "صفر";
  if (n < 0) return `سالب ${integerToArabicWords(-n)}`;

  const parts: string[] = [];
  let rem = n;

  for (const scale of SCALES) {
    if (rem < scale.value) continue;
    const count = Math.floor(rem / scale.value);
    rem %= scale.value;
    parts.push(scaleLabel(count, scale));
  }

  if (rem > 0) {
    parts.push(underThousand(rem));
  }

  return parts.join(" و");
}

function parseAmountNumber(raw: string): number | null {
  const cleaned = raw.replace(/,/g, "").trim();
  if (!cleaned) return null;
  const n = Number.parseFloat(cleaned);
  if (!Number.isFinite(n)) return null;
  return n;
}

/** يعرض المبلغ كتابةً بالريال (والهللة إن وُجدت كسور). */
export function amountToArabicWords(raw: string | number): string {
  const n = typeof raw === "number" ? raw : parseAmountNumber(raw);
  if (n == null) return "—";
  if (n === 0) return "صفر ريال سعودي";

  const negative = n < 0;
  const abs = Math.abs(n);
  const riyals = Math.floor(abs + 1e-9);
  const hallalas = Math.round((abs - riyals) * 100);

  const parts: string[] = [];
  if (riyals > 0) {
    parts.push(`${integerToArabicWords(riyals)} ريال سعودي`);
  } else if (hallalas === 0) {
    parts.push("صفر ريال سعودي");
  }

  if (hallalas > 0) {
    parts.push(`${integerToArabicWords(hallalas)} هللة`);
  }

  const text = parts.join(" و");
  return negative ? `سالب ${text}` : text;
}

export function formatAmountNumberDisplay(raw: string | number): string {
  const n = typeof raw === "number" ? raw : parseAmountNumber(raw);
  if (n == null) return "—";
  return new Intl.NumberFormat("en-US", {
    maximumFractionDigits: 2,
    minimumFractionDigits: Number.isInteger(n) ? 0 : 2,
  }).format(n);
}

File: backend/contexts/valuation/RealEstateEval.Valuation.Domain/ArabicAmountWords.cs
namespace RealEstateEval.Valuation.Domain;

/// <summary>Arabic amount words (تفقيط) for report display and field code 65167.</summary>
public static class ArabicAmountWords
{
    private static readonly string[] Ones =
    [
        "", "واحد", "اثنان", "ثلاثة", "أربعة", "خمسة", "ستة", "سبعة", "ثمانية", "تسعة",
        "عشرة", "أحد عشر", "اثنا عشر", "ثلاثة عشر", "أربعة عشر", "خمسة عشر",
        "ستة عشر", "سبعة عشر", "ثمانية عشر", "تسعة عشر",
    ];

    private static readonly string[] Tens =
        ["", "", "عشرون", "ثلاثون", "أربعون", "خمسون", "ستون", "سبعون", "ثمانون", "تسعون"];

    private static readonly string[] Hundreds =
    [
        "", "مائة", "مائتان", "ثلاثمائة", "أربعمائة", "خمسمائة",
        "ستمائة", "سبعمائة", "ثمانمائة", "تسعمائة",
    ];

    public static string AmountToArabicWords(decimal amount)
    {
        if (amount == 0m) return "صفر ريال سعودي";

        var negative = amount < 0m;
        var abs = Math.Abs(amount);
        var riyals = (long)Math.Floor(abs + 0.000000001m);
        var hallalas = (int)Math.Round((abs - riyals) * 100m, MidpointRounding.AwayFromZero);
        if (hallalas == 100)
        {
            riyals += 1;
            hallalas = 0;
        }

        var parts = new List<string>();
        if (riyals > 0)
            parts.Add($"{IntegerToArabicWords(riyals)} ريال سعودي");
        else if (hallalas == 0)
            parts.Add("صفر ريال سعودي");

        if (hallalas > 0)
            parts.Add($"{IntegerToArabicWords(hallalas)} هللة");

        var text = string.Join(" و", parts);
        return negative ? $"سالب {text}" : text;
    }

    public static string IntegerToArabicWords(long n)
    {
        if (n == 0) return "صفر";
        if (n < 0) return $"سالب {IntegerToArabicWords(-n)}";

        var parts = new List<string>();
        long rem = n;

        AppendScale(parts, ref rem, 1_000_000_000, "مليار", "ملياران", "مليارات", "مليار");
        AppendScale(parts, ref rem, 1_000_000, "مليون", "مليونان", "ملايين", "مليون");
        AppendScale(parts, ref rem, 1_000, "ألف", "ألفان", "آلاف", "ألف");

        if (rem > 0)
            parts.Add(UnderThousand((int)rem));

        return string.Join(" و", parts);
    }

    private static void AppendScale(
        List<string> parts,
        ref long rem,
        long value,
        string one,
        string dual,
        string few,
        string many)
    {
        if (rem < value) return;
        var count = rem / value;
        rem %= value;
        parts.Add(ScaleLabel(count, one, dual, few, many));
    }

    private static string ScaleLabel(long count, string one, string dual, string few, string many)
    {
        if (count == 1) return one;
        if (count == 2) return dual;
        if (count is >= 3 and <= 10)
            return $"{UnderThousand((int)count)} {few}";
        return $"{UnderThousand((int)count)} {many}";
    }

    private static string UnderThousand(int n)
    {
        if (n < 100) return UnderHundred(n);
        var h = n / 100;
        var rest = n % 100;
        var head = Hundreds[h];
        if (rest == 0) return head;
        return $"{head} و{UnderHundred(rest)}";
    }

    private static string UnderHundred(int n)
    {
        if (n < 20) return Ones[n];
        var t = n / 10;
        var o = n % 10;
        if (o == 0) return Tens[t];
        return $"{Ones[o]} و{Tens[t]}";
    }
}
